File: Reputex-GeminiBackend/app/models/business.py
"""Business and BrandKeyword models."""
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional
from sqlalchemy import Boolean, DateTime, ForeignKey, JSON, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.models.base import Base, UUIDMixin, utc_now


class Business(Base, UUIDMixin):
    __tablename__ = "businesses"

    name: Mapped[str] = mapped_column(String(255), index=True, nullable=False)
    category: Mapped[str] = mapped_column(String(100), default="General", nullable=False)
    website: Mapped[Optional[str]] = mapped_column(String(500), nullable=True)
    location: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    phone: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)
    description: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    monitored_platforms: Mapped[List[str]] = mapped_column(JSON, default=list, nullable=False)
    owner_id: Mapped[Optional[str]] = mapped_column(String(36), index=True, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utc_now, nullable=False)
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utc_now, onupdate=utc_now, nullable=False)

    keywords: Mapped[List["BrandKeyword"]] = relationship("BrandKeyword", back_populates="business", cascade="all, delete-orphan")


class BrandKeyword(Base, UUIDMixin):
    __tablename__ = "brand_keywords"

    business_id: Mapped[str] = mapped_column(String(36), ForeignKey("businesses.id", ondelete="CASCADE"), index=True, nullable=False)
    keyword: Mapped[str] = mapped_column(String(255), nullable=False)
    category: Mapped[str] = mapped_column(String(50), default="brand", nullable=False)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=utc_now, nullable=False)

    business: Mapped["Business"] = relationship("Business", back_populates="keywords")

File: Reputex-App/lib/features/onboarding/domain/models/business.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'brand_keyword.dart';

part 'business.freezed.dart';
part 'business.g.dart';

@freezed
abstract class Business with _$Business {
  const factory Business({
    required String id,
    required String name,
    required String category,
    String? website,
    String? location,
    String? phone,
    @JsonKey(name: 'monitored_platforms')
    @Default([])
    List<String> monitoredPlatforms,
    @Default([]) List<BrandKeyword> keywords,
    @JsonKey(name: 'owner_id') String? ownerId,
    @JsonKey(name: 'created_at') DateTime? createdAt,
  }) = _Business;

  factory Business.fromJson(Map<String, dynamic> json) =>
      _$BusinessFromJson(json);
}
